### ISSUE TITLE
Counterclockwise code partitions / folding

### ISSUE THREAD

**Issue by arichiardi** [laurentpetit/ccw#810]
I wanted to open an issue to ptrack ideas about code partitioning and folding in Counterclockwise.

JDT's way is complicated, basically they implemented pluggable folding providers through extensions I guess to be able to add (even at runtime) new policies. It is basically what I have done with hovers, but I don't see a big advantage to have such flexibility here (I might be of course super wrong).

The plan would be to implement:
- [ ] Partitioning
- [ ] Folding

Random thoughts:
- Probably ClojureTokenScanner can be re-used for the first part, but I have to dig into it.
- Partitions should not overlap. therefore I already see a problem when trying to define a partition for top level forms and their embedded docstring.

In the meantime, some info that I found:
[Documents and Partitions](http://help.eclipse.org/luna/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Feditors_documents.htm)
[Partitions are non-overlapping](https://wiki.eclipse.org/FAQ_What_is_a_document_partition%3F)
[Eclipse strightforward folding](https://eclipse.org/articles/Article-Folding-in-Eclipse-Text-Editors/folding.html)

**Comment by arichiardi** [comment]
Ah, partitioning is already implemented it seems, I would like to have some confirmation Laurent, because ```ClojureDocumentProvider``` is in place...

**Comment by arichiardi** [comment]
Ok I have noticed that just a few partitions are implemented and partially used at the moment. Taking inspiration from JDT as usual, I would like to write a Partition Scanner that is more than Rule based (if worth it). I am going to need some hammock time now :)

**Comment by laurentpetit** [comment]
Yes, I also always asked myself how partitions are meant to be used. They don't have to overlap. So for instance the JDT, as far as I remember, has partitions for "code", "comments" and "strings". I see this as a way to plugin different code completions, syntax coloring strategies, folding strategies, when inside "raw code", "comments" or "strings".
It also looks like a way to separate the code in regions, with the hope that the split is somehow homogeneous, and that it will help with performance when reconciling models while the user is typing.

I also read that it's now possible to have different "partionings", e.g. each partitioning decomposes the editor content differently. I understand that this way the partitions relevant for code completion can be different than the partitions relevant for syntax highlighting, etc.

I have always asked myself for which parts of ccw we need partitions. Except for reusing the "folding" feature provided by Eclipse out-of-the-box, I don't see much use yet.

**Comment by arichiardi** [comment]
Yes I saw the multiple partitioning stuff...but I still need to understand where to plug it...in our DocumentProvider I added something but still no brilliant results.

I was as puzzled as you by the way, especially because partitions cannot be nested...

One nice thing for ccw is if we can distinguish (and we can) between a string and a docstring...the formatting of a docstring can be set as we like then (or set through preference indipindently).

Another use that I had in mind... I adapted clojure.core's reader code for intercepting char literals and we could aid completion/syntax validation of all the literals as well through partitions....

Still experimenting but I see some light far far away :)

**Comment by laurentpetit** [comment]
Yes, with ccw and the parsley incremental parser, we already have, always in sync, the parse tree. 

And we can easily add support for partitions, or any kind of semantic views over the edited code, by adding views. It's all located inside the paredit parser namespace. Working on cleaning the code is still a TBD task, however, and it's probably hard to grasp right now without support from me or Christophe. 




I don't understand why/how we need support for partitions or a dedicated partitioning for enhancing code assist (code completion) for char litterals?



—
Envoyé avec Mailbox

**Comment by arichiardi** [comment]
No no I just enchanted the existing partition scheme that was not doing much at the moment...I will open a PR when done.. But thanks for the hint to parsley...I will look into it :smile:

**Comment by arichiardi** [comment]
I am checking the parse tree at the moment as I will need it for sure...I have one question though that I figure I might ask you...

The burden of the creation of the parse tree is in the ClojureSourceViewer, which, being a viewer, is initialized long after the Document partitioners...

In particular the init occurs (it seems to me) in:
```Java
private void updateTextBuffer (String finalText, long offset, long length, String text) {
    	boolean firstTime = (parseState == null);
    	parseState = editorSupport._("updateTextBuffer",parseState, finalText, offset, length, text);
        if (firstTime) {
        	editorSupport._("startWatchParseRef", parseState, this);
        }
    }
```

Now, I know that you don't like huge refactorings (:devil:) but in order to provide an alternative partitioning based on it, I would need to create the tree as soon as the ```ClojureEditor``` is up...and it makes sense...

Or...if you have any other advice on how to approach this.

**Comment by arichiardi** [comment]
I saw a comment calling for a ```ClojureDocument```...maybe keeping the tree in there is the best idea.

**Comment by laurentpetit** [comment]
Wait. I remember Christophe Grand had some work he wanted to merge, related to refactoring of the editor document, etc. @cgrand do you confirm?



—
Envoyé avec Mailbox

**Comment by arichiardi** [comment]
Ok...You mean the other PR? no it is not Eclipse-related, but a ```parsley``` update. I would say that folding is not possible without changing these things a bit, I will wait...I can also describe you what are my discoveries...

**Comment by arichiardi** [comment]
In any case, I think I will manage, at least in the first iteration, to not change too much...so no huge refactoring (I hope) will occurr.

**Comment by arichiardi** [comment]
In PR #846
